_posts/2015-12-14-simply-secure-rails-api.markdown

```diff
@@ -0,0 +1,733 @@
+                      faraday.adapter Faraday.default_adapter    
+                      faraday.headers['Content-Type'] = 'application/json'
+                    end
+
+      response = connection.post("/api/v1/login/", {
+                    email: email,
+                    password: password
+                  }.to_json)
+
+      parsed_body = JSON.parse(response.body)
+
+      user = User.create({
+        uuid:  parsed_body["user"]["uuid"],
+        email: parsed_body["user"]["email"],
+        name:  parsed_body["user"]["name"]
+      })
+
+      {user: user, payload: parsed_body}
+    rescue 
+      false
+    end
+  end
+end
+```
+
+Here we're creating a random token, per user, which we can alter / delete if the top level ever becomes compromised.
+
+
+##### Our modified Social Network login view
+```html
+<html>
+  <head>
+    <title>Network</title>
+  </head>
+  <body>
+    <form method="POST" action="<%= create_login_path %>" id="loginForm">
+      <input type="text" name="email" placeholder="Email" />
+      <input type="password" name="password" placeholder="password" />
+      <input type="submit" value="submit" />
+    </form>
+    <script>
+      $("#loginForm").on("submit", function(){
+        $.ajax({
+          url: $(this).attr("action"),
+          method: "POST",
+          data: $(this).serialize()
+        })
+        .success(function(response) {
+          localStorage.setItem("e", response.payload.encrypted);
+          localStorage.setItem("t", response.user.token);
+        });
+      });
+    </script>
+  </body>
+</html>
+```
+
+##### Our new Social Network's Application Controller
+
+```ruby
+class ApplicationController < ActionController::Base
+
+  protected
+
+  def decrypt(encrypted_uuid)
+    begin
+      decipher         = OpenSSL::Cipher.new 'aes-256-cbc'
+      decipher.decrypt
+      decipher.padding = 1
+      decipher.key     = ENV['SOME_SECRET_KEY']
+      decipher.iv      = ENV['SOME_SECRET_IV']
+      decrypted        = cipher.update(Base64.urlsafe_decode64(encrypted_uuid)) + decipher.final
+      decrypted
+    rescue
+      head 403 and return
+    end
+  end
+
+  def current_user
+    if params[:encrypted_user_uuid] && params[:token]
+      @current_user ||= User.find_by(uuid: decrypt_uuid(params[:encrypted_user_uuid]), token: params[:token])
+    end
+  end
+  helper_method :current_user
+
+  def require_user
+    if !current_user.present?
+      head 403
+    end
+  end
+
+end
+```
+
+Now we are accessing this user based on an easily revoke-able token in the client app.
+
+---
+
+##### Problem 2: Decoding
+
+Since we are encoding a uuid and the encrypted uuid is stored in the client its *technically possible* ([the best kind of possible](https://www.youtube.com/watch?v=hou0lU8WMgo)) for a nefarious person to gather up 
+enough uuids and their encrypted counterparts to break our encryption. 
+
+So lets add some additional values to our encryption for parsing.
+
```

Comment:
Would be interesting to discuss the order of magnitude of how many uuids you would need to do this. If it's outside actual practical bounds then you could dismiss this section. Or should we actually recommend it?